de.metas.business/src/main/java/de/metas/customs/process/M_InOut_Create_CustomsInvoice.java

```diff
@@ -0,0 +1,224 @@
+				.map(InOutAndLineId::getInOutId)
+				.collect(ImmutableSet.toImmutableSet());
+
+		customsInvoiceService.setCustomsInvoiceToShipments(exportedShippmentIds, customsInvoice);
+
+		return MSG_OK;
+
+	}
+
+	private ProductId getProductId(final InOutAndLineId inoutAndLineId)
+	{
+		final IInOutDAO inOutDAO = Services.get(IInOutDAO.class);
+
+		final InOutLineId shipmentLineId = inoutAndLineId.getInOutLineId();
+		final I_M_InOutLine shipmentLineRecord = inOutDAO.getLineById(shipmentLineId, I_M_InOutLine.class);
+
+		return ProductId.ofRepoId(shipmentLineRecord.getM_Product_ID());
+	}
+
+	@Override
+	public ProcessPreconditionsResolution checkPreconditionsApplicable(@NonNull final IProcessPreconditionsContext context)
+	{
+
+		if (context.getSelectionSize() <= 0)
+		{
+			return ProcessPreconditionsResolution.rejectBecauseNoSelection();
+		}
+
+		final boolean foundAtLeastOneUnregisteredShipment = context.getSelectedModels(I_M_InOut.class).stream()
+				.map(shipmentRecord -> InOutId.ofRepoId(shipmentRecord.getM_InOut_ID()))
+				.filter(this::isValidShipment)
+				.map(inoutId -> retrieveValidLinesToExport(inoutId))
+				.findAny()
+				.isPresent();
+
+		return ProcessPreconditionsResolution.acceptIf(foundAtLeastOneUnregisteredShipment);
+	}
+
+	private boolean isValidShipment(final InOutId shipmentId)
+	{
+		final IDocumentBL documentBL = Services.get(IDocumentBL.class);
+
+		final IInOutBL inOutBL = Services.get(IInOutBL.class);
+		final IInOutDAO inOutDAO = Services.get(IInOutDAO.class);
+
+		final I_M_InOut shipment = inOutDAO.getById(shipmentId);
+
+		if (!shipment.isSOTrx())
+		{
+			return false;
+		}
+
+		if (inOutBL.isReversal(shipment))
+		{
+			return false;
+		}
+
+		if (!documentBL.isStatusCompletedOrClosedOrReversed(shipment.getDocStatus()))
+		{
+			return false;
+		}
+
+		if (shipment.isExportedToCustomsInvoice())
+		{
+			return false;
+		}
+
+		return true;
+	}
+
+	private List<InOutAndLineId> retrieveLinesToExport()
+	{
+		final IQueryFilter<I_M_InOut> queryFilter = getProcessInfo()
+				.getQueryFilterOrElse(ConstantQueryFilter.of(false));
+
+		final ImmutableList<InOutId> shipmentsToExport = Services.get(IQueryBL.class)
+				.createQueryBuilder(I_M_InOut.class)
+				.filter(queryFilter)
+				.create()
+				.listIds(InOutId::ofRepoId)
+				.stream()
+				.filter(this::isValidShipment)
+				.collect(ImmutableList.toImmutableList());
+
+		List<InOutAndLineId> shipmentLinesToExport = new ArrayList<>();
+
+		shipmentsToExport
+				.stream()
+				.forEach(shipmentId -> shipmentLinesToExport.addAll(retrieveValidLinesToExport(shipmentId)));
+
+		return shipmentLinesToExport;
+
+	}
+
+	private List<InOutAndLineId> retrieveValidLinesToExport(final InOutId shipmentId)
+	{
+		final IInOutDAO inOutDAO = Services.get(IInOutDAO.class);
+		List<InOutAndLineId> shipmentLines = new ArrayList<>();
+
+		final ImmutableList<InOutAndLineId> shipmentLinesToExport = inOutDAO.retrieveLinesForInOutId(shipmentId)
```

Comment:
* if the service class only supports shipmentline with an orderline, i think you should filter for that in here
* maybe instead of filtering with a predicate, you should create some service/DAO/repository to get those lines, and in there, use IQuery to just get from the DB the stuff you really need

Reply:
done

Reply:
sry, i had in mind, to filter for it when making the IQuery. 
but i guess we won't have 1000s of shipmentlines without orderlines, so this is also OK

Reply:
if all goes well, all the lines with no C_OrderLine_ID will already be dropped at this point: 
```
	if (shipmentLineRecord.isPackagingMaterial())
		{
			return false;
		}

```